Fix Cuda IPC deadlock

VitalyFedyunin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#40347 Fix Cuda IPC deadlock**

Fixes: #39541 
Fixes: #25301

Differential Revision: [D22152662](https://our.internmc.facebook.com/intern/diff/D22152662)

VitalyFedyunin committed:
Fix Cuda IPC deadlock

[ghstack-poisoned]

VitalyFedyunin commented on torch/csrc/CudaIPCTypes.cpp:
```diff
@@ -61,17 +61,23 @@ CudaIPCSentDataLimbo::~CudaIPCSentDataLimbo() {
     }
+    shared_blocks_ = std::move(kept_blocks);
+  }
+  for (auto& sd : reset_blocks) {
```
Need to reset blocks out of the critical section here, otherwise it deadlocks.

ngimel replied:
add it as a comment to the code

ngimel commented on test/test_multiprocessing.py:
```diff
@@ -37,6 +38,29 @@ def run(self):
 
 
+def _test_cuda_ipc_deadlock_actor(queue):
+    steps = 0
```
why not loop over steps?
Also, better to send `steps` as argument instead of hardcoding it in 2 functions - they'll deadlock if number of steps is not the same

ngimel reviewed (approved):
Looks good. Please link to a recent issue that this is fixing, and older issues if any.

VitalyFedyunin committed:
Update per PR comments on "Fix Cuda IPC deadlock"


Differential Revision: [D22152662](https://our.internmc.facebook.com/intern/diff/D22152662)

[ghstack-poisoned]

VitalyFedyunin committed:
Linter fix on "Fix Cuda IPC deadlock"


Fixes: #39541 
Fixes: #25301

Differential Revision: [D22152662](https://our.internmc.facebook.com/intern/diff/D22152662)

[ghstack-poisoned]

VitalyFedyunin committed:
rebase on "Fix Cuda IPC deadlock"


Fixes: #39541 
Fixes: #25301

Differential Revision: [D22152662](https://our.internmc.facebook.com/intern/diff/D22152662)

[ghstack-poisoned]
